File: src/decode.ts
/**
 * ChromaCode decoder: PNG image → data.
 *
 * Pipeline:
 * 1. Read PNG → pixel buffer
 * 2. Detect quiet zone / cell size from finder patterns
 * 3. Extract grid dimensions
 * 4. Read header from first N data cells
 * 5. Extract data cells in serpentine order
 * 6. Decode channels → interleaved stream
 * 7. Deinterleave → RS blocks
 * 8. RS decode each block (correct errors)
 * 9. Concatenate data
 * 10. Optional inflate decompression
 */

import { inflateSync } from 'node:zlib'
import type { EncodingMode } from './types.js'
import { BYTES_PER_CELL, EC_RATIO, RS_MAX_BLOCK, QUIET_ZONE } from './types.js'
import { rsDecode, ecSymbolCount } from './reed-solomon.js'
import { deinterleave } from './interleave.js'
import { allocateGrid, decodeHeader, headerSize } from './grid.js'
import { decodeCell } from './channels.js'
import { readPng } from './png.js'

/**
 * Decode a ChromaCode PNG image back to binary data.
 */
export function decode(png: Uint8Array): Uint8Array {
  // Step 1: Read PNG
  const image = readPng(png)
  const { pixels, width: pixelWidth, height: pixelHeight, bitDepth } = image
  const bpp = bitDepth === 16 ? 8 : 4

  // Step 2: Detect cell size by finding the top-left finder pattern
  // The TL finder starts at (QUIET_ZONE * cellSize, QUIET_ZONE * cellSize)
  // in pixel coordinates. We detect it by scanning for the finder's outer color.
  const cellSize = detectCellSize(pixels, pixelWidth, pixelHeight, bitDepth)

  // Step 3: Extract grid dimensions from pixel dimensions
  const gridWidth = Math.floor(pixelWidth / cellSize) - 2 * QUIET_ZONE
  const gridHeight = Math.floor(pixelHeight / cellSize) - 2 * QUIET_ZONE

  if (gridWidth < 12 || gridHeight < 12) {
    throw new Error(`Grid too small: ${gridWidth}×${gridHeight}`)
  }

  // Helper: read a cell's pixel value (top-left pixel of the cell)
  const readCell = (cellX: number, cellY: number): Uint8Array => {
    const px = (cellX + QUIET_ZONE) * cellSize
    const py = (cellY + QUIET_ZONE) * cellSize
    const offset = (py * pixelWidth + px) * bpp
    return pixels.subarray(offset, offset + bpp)
  }

  // Step 4: Allocate grid to get data cell coordinates
  const grid = allocateGrid(gridWidth, gridHeight)

  // Read header: first N data cells
  // We need to read enough cells to get the header, which tells us the mode.
  // Start by trying the maximum header size (18 bytes / smallest cell = 18/3 = 6 cells)
  // We'll read in rgba64 mode first (8 bytes/cell), then re-read if needed.

  // Actually, we can determine the bit depth from the PNG itself.
  // The mode is encoded in the header. We need to bootstrap:
  // Read enough cells to get the first header byte (which contains version + mode).
  // Then we know the bytes per cell and can read the full header.

  // For any mode, the first header byte is at offset 0 in the first cell's data.
  // Read the first cell with the most conservative mode (smallest bytes per cell).
  // First cell's first byte always contains version(4b) | mode(4b).

  // Read the first data cell raw pixels
  const firstCellPixel = readCell(...grid.dataCoords[0])

  // Determine mode from first byte: try all modes to read byte 0
  // In all modes, the first data byte maps to the red channel high byte (16-bit) or red byte (8-bit)
  let modeNibble: number
  if (bitDepth === 16) {
    // First pixel R channel: high byte of uint16
    const view = new DataView(
      firstCellPixel.buffer,
      firstCellPixel.byteOffset,
      firstCellPixel.byteLength,
    )
    const r16 = view.getUint16(0)
    modeNibble = r16 >> 8 // First byte of data = high byte of R16
  } else {
    // 8-bit: first byte = R channel
    modeNibble = firstCellPixel[0]
  }

  // Extract mode from the first header byte
  const modeId = modeNibble & 0x0f
  const modeMap: Record<number, EncodingMode> = {
    0: 'rgba64',
    1: 'rgba32',
    2: 'rgb48',
    3: 'rgb24',
  }
  const mode = modeMap[modeId]
  if (!mode) throw new Error(`Unknown mode ID in header: ${modeId}`)

  const bytesPerCell = BYTES_PER_CELL[mode]
  const hasSeqByte1 = readHeaderByte1(grid, readCell, mode, bitDepth)
  const hasSequence = (hasSeqByte1 & 0x20) !== 0
  const hdrSize = headerSize(hasSequence)
  const headerCells = Math.ceil(hdrSize / bytesPerCell)

  // Read full header
  const headerBytes = new Uint8Array(headerCells * bytesPerCell)
  for (let i = 0; i < headerCells; i++) {
    const [cx, cy] = grid.dataCoords[i]
    const pixel = readCell(cx, cy)
    const cellData = decodeCell(pixel, mode)
    headerBytes.set(cellData, i * bytesPerCell)
  }

  const header = decodeHeader(headerBytes.subarray(0, hdrSize))

  // Verify grid dimensions match
  if (header.gridWidth !== gridWidth || header.gridHeight !== gridHeight) {
    throw new Error(
      `Grid dimension mismatch: header says ${header.gridWidth}×${header.gridHeight}, ` +
        `image gives ${gridWidth}×${gridHeight}`,
    )
  }

  // Step 5-6: Read data cells and decode
  const dataCoords = grid.dataCoords.slice(headerCells)
  const payloadLen = header.dataLength
  const ecRatio = EC_RATIO[header.ecLevel]

  // Reconstruct block structure from payload length
  let dataPerBlock: number
  let ecCount: number

  if (payloadLen === 0) {
    // Empty data case
    return new Uint8Array(0)
  }

  // Find the same data-per-block as encoder
  for (dataPerBlock = RS_MAX_BLOCK - 2; dataPerBlock >= 1; dataPerBlock--) {
    ecCount = ecSymbolCount(dataPerBlock, ecRatio)
    if (dataPerBlock + ecCount <= RS_MAX_BLOCK) break
  }
  const numBlocks = Math.ceil(payloadLen / dataPerBlock!)
  const blockSizes: number[] = []

  for (let i = 0; i < numBlocks; i++) {
    const start = i * dataPerBlock!
    const end = Math.min(start + dataPerBlock!, payloadLen)
    const blockDataLen = end - start
    const blockEcCount = ecSymbolCount(blockDataLen, ecRatio)
    blockSizes.push(blockDataLen + blockEcCount)
  }

  const totalStreamBytes = blockSizes.reduce((s, b) => s + b, 0)
  const totalStreamCells = Math.ceil(totalStreamBytes / bytesPerCell)

  // Read interleaved stream from data cells
  const rawStream = new Uint8Array(totalStreamCells * bytesPerCell)
  for (let i = 0; i < totalStreamCells && i < dataCoords.length; i++) {
    const [cx, cy] = dataCoords[i]
    const pixel = readCell(cx, cy)
    const cellData = decodeCell(pixel, mode)
    rawStream.set(cellData, i * bytesPerCell)
  }

  const interleavedStream = rawStream.subarray(0, totalStreamBytes)

  // Step 7: Deinterleave
  const rsBlocks = deinterleave(interleavedStream, blockSizes)

  // Step 8: RS decode each block
  const dataChunks: Uint8Array[] = []
  for (let i = 0; i < numBlocks; i++) {
    const blockDataLen = Math.min(dataPerBlock!, payloadLen - i * dataPerBlock!)
    const blockEcCount = rsBlocks[i].length - blockDataLen
    const decoded = rsDecode(rsBlocks[i], blockEcCount)
    dataChunks.push(decoded)
  }

  // Step 9: Concatenate
  const totalData = dataChunks.reduce((s, c) => s + c.length, 0)
  const result = new Uint8Array(totalData)
  let offset = 0
  for (const chunk of dataChunks) {
    result.set(chunk, offset)
    offset += chunk.length
  }

  // Truncate to exact data length
  const trimmed = result.subarray(0, payloadLen)

  // Step 10: Decompress if needed
  if (header.compressed) {
    return new Uint8Array(inflateSync(Buffer.from(trimmed)))
  }

  return trimmed
}

/**
 * Read header byte 1 (contains hasSequence flag) by reading the second byte
 * from the appropriate cell.
 */
function readHeaderByte1(
  grid: ReturnType<typeof allocateGrid>,
  readCell: (cx: number, cy: number) => Uint8Array,
  mode: EncodingMode,
  _bitDepth: 8 | 16,
): number {
  const bytesPerCell = BYTES_PER_CELL[mode]

  if (bytesPerCell >= 2) {
    // Byte 1 is in the first cell
    const [cx, cy] = grid.dataCoords[0]
    const pixel = readCell(cx, cy)
    const cellData = decodeCell(pixel, mode)
    return cellData[1]
  } else {
    // Byte 1 is in the second cell (only for hypothetical 1-byte/cell modes, which we don't have)
    const [cx, cy] = grid.dataCoords[1]
    const pixel = readCell(cx, cy)
    const cellData = decodeCell(pixel, mode)
    return cellData[0]
  }
}

/**
 * Detect cell size by examining the top-left finder pattern.
 * Scans from the quiet zone boundary to find the first finder-colored pixel,
 * then measures the cell size.
 */
function detectCellSize(
  pixels: Uint8Array,
  pixelWidth: number,
  _pixelHeight: number,
  bitDepth: 8 | 16,
): number {
  const bpp = bitDepth === 16 ? 8 : 4

  // The finder starts at pixel (QUIET_ZONE * cellSize, QUIET_ZONE * cellSize).
  // We know QUIET_ZONE = 1, so the finder starts at (cellSize, cellSize).
  // Before that is the quiet zone (transparent).
  // Scan row 0 to find where the quiet zone ends (first non-zero alpha)
  // then that's the cellSize boundary.

  // Actually, scan the first row. The quiet zone has alpha=0, then at x=cellSize
  // the finder starts with the outer color (full alpha).
  // So find the first pixel with non-zero alpha.

  // For cell size detection: scan along y=0 row.
  // Quiet zone pixels have alpha=0, finder pixels have full alpha.
  // The first pixel with full alpha is at x = QUIET_ZONE * cellSize.

  // Then to determine cellSize: the outer color of the finder continues for
  // cellSize pixels, then transitions to white (layer 1) at x = (QUIET_ZONE+1)*cellSize.

  // First, find start of finder (first non-transparent pixel in row y)
  // Try multiple rows for robustness
  for (let testY = 0; testY < Math.min(50, _pixelHeight); testY++) {
    let finderStart = -1

    for (let x = 0; x < pixelWidth; x++) {
      const offset = (testY * pixelWidth + x) * bpp
      const alpha =
        bitDepth === 16 ? (pixels[offset + 6] << 8) | pixels[offset + 7] : pixels[offset + 3]

      if (alpha > 0) {
        finderStart = x
        break
      }
    }

    if (finderStart < 0) continue

    // cellSize = finderStart / QUIET_ZONE
    const cellSize = Math.round(finderStart / QUIET_ZONE)
    if (cellSize < 1) continue

    // Verify: at (finderStart, testY) we should see the outer finder color
    // and at (finderStart + cellSize, testY + cellSize) we should see white (if in finder range)
    return cellSize
  }

  // Fallback: assume cellSize = 1
  return 1
}

File: native/src/decode.c
/*
 * decode.c — ChromaCode decoder: PNG image → data.
 *
 * Pipeline:
 * 1. Read PNG → pixel buffer
 * 2. Detect cell size from finder patterns
 * 3. Extract grid dimensions
 * 4. Read header from first N data cells
 * 5. Extract data cells in serpentine order
 * 6. Decode channels → interleaved stream
 * 7. Deinterleave → RS blocks
 * 8. RS decode each block (correct errors)
 * 9. Concatenate data
 * 10. Optional inflate decompression
 */

#define CHROMACODE_INTERNAL
#include "chromacode.h"
#include <stdlib.h>
#include <string.h>
#include <zlib.h>

/* ── Helper: read cell pixel from pixel buffer ───────────────────── */

static const uint8_t *read_cell(const uint8_t *pixels, int pixel_width,
                                 int bpp, int cell_size,
                                 int cell_x, int cell_y)
{
    int px = (cell_x + CC_QUIET_ZONE) * cell_size;
    int py = (cell_y + CC_QUIET_ZONE) * cell_size;
    size_t offset = ((size_t)py * pixel_width + px) * (size_t)bpp;
    return pixels + offset;
}

/* ── Detect cell size ────────────────────────────────────────────── */

static int detect_cell_size(const uint8_t *pixels, int pixel_width,
                            int pixel_height, int bit_depth)
{
    int bpp = (bit_depth == 16) ? 8 : 4;

    /* Scan rows looking for the first non-transparent pixel.
     * Quiet zone has alpha=0. The finder starts at pixel (cellSize, cellSize). */
    for (int y = 0; y < pixel_height && y < 50; y++) {
        for (int x = 0; x < pixel_width; x++) {
            size_t offset = ((size_t)y * pixel_width + x) * (size_t)bpp;
            int alpha;
            if (bit_depth == 16) {
                alpha = ((int)pixels[offset + 6] << 8) | pixels[offset + 7];
            } else {
                alpha = pixels[offset + 3];
            }

            if (alpha > 0) {
                /* cellSize = x / QUIET_ZONE */
                int cell_size = x / CC_QUIET_ZONE;
                if (cell_size < 1) cell_size = 1;
                /* Round: in TS it's Math.round(x / QUIET_ZONE) */
                return cell_size;
            }
        }
    }

    return 1; /* fallback */
}

/* ── Decode ──────────────────────────────────────────────────────── */

int cc_decode(const uint8_t *png, size_t png_len,
              uint8_t **out_data, size_t *out_len)
{
    cc_gf_init();

    /* Step 1: Read PNG */
    uint8_t *pixels = NULL;
    int pixel_width = 0, pixel_height = 0, bit_depth = 0;
    int rc = cc_png_read(png, png_len, &pixels, &pixel_width, &pixel_height,
                         &bit_depth);
    if (rc != CC_OK) return rc;

    int bpp = (bit_depth == 16) ? 8 : 4;

    /* Step 2: Detect cell size */
    int cell_size = detect_cell_size(pixels, pixel_width, pixel_height,
                                     bit_depth);

    /* Step 3: Grid dimensions */
    int grid_w = pixel_width / cell_size - 2 * CC_QUIET_ZONE;
    int grid_h = pixel_height / cell_size - 2 * CC_QUIET_ZONE;

    if (grid_w < 12 || grid_h < 12) {
        free(pixels);
        return CC_ERR_DECODE;
    }

    /* Step 4: Allocate grid for data coords */
    cc_grid grid;
    rc = cc_grid_alloc(grid_w, grid_h, &grid);
    if (rc != CC_OK) { free(pixels); return rc; }

    /* Bootstrap: read first cell to get mode */
    const uint8_t *first_px = read_cell(pixels, pixel_width, bpp, cell_size,
                                         grid.data_coords[0],
                                         grid.data_coords[1]);

    int mode_nibble = first_px[0];

    int mode_id = mode_nibble & 0x0F;
    if (mode_id < 0 || mode_id > 3) {
        cc_grid_free(&grid);
        free(pixels);
        return CC_ERR_HEADER;
    }

    int mode = mode_id;
    int bytes_per_cell = CC_BYTES_PER_CELL[mode];
    /* ec_ratio_100 set after header decode below */
    int ec_ratio_100;

    /* Read byte 1 to get has_sequence flag */
    uint8_t cell0_data[8];
    cc_decode_cell(first_px, mode, cell0_data);
    int has_sequence = (cell0_data[1] & 0x20) ? 1 : 0;
    int hdr_size = cc_header_size(has_sequence);
    int header_cells = (hdr_size + bytes_per_cell - 1) / bytes_per_cell;

    /* Read full header */
    int hdr_buf_len = header_cells * bytes_per_cell;
    uint8_t *hdr_buf = (uint8_t *)calloc((size_t)hdr_buf_len, 1);
    if (!hdr_buf) { cc_grid_free(&grid); free(pixels); return CC_ERR_ALLOC; }

    for (int i = 0; i < header_cells; i++) {
        int cx = grid.data_coords[i * 2];
        int cy = grid.data_coords[i * 2 + 1];
        const uint8_t *px = read_cell(pixels, pixel_width, bpp, cell_size,
                                       cx, cy);
        cc_decode_cell(px, mode, hdr_buf + i * bytes_per_cell);
    }

    cc_header header;
    rc = cc_decode_header(hdr_buf, hdr_buf_len, &header);
    free(hdr_buf);
    if (rc != CC_OK) { cc_grid_free(&grid); free(pixels); return rc; }

    /* Verify grid dimensions */
    if (header.grid_width != grid_w || header.grid_height != grid_h) {
        cc_grid_free(&grid);
        free(pixels);
        return CC_ERR_HEADER;
    }

    uint32_t payload_len = header.data_length;
    if (header.ec_level < 0 || header.ec_level > 3) {
        cc_grid_free(&grid);
        free(pixels);
        return CC_ERR_HEADER;
    }
    ec_ratio_100 = CC_EC_RATIO_100[header.ec_level];

    /* Empty data */
    if (payload_len == 0) {
        cc_grid_free(&grid);
        free(pixels);
        *out_data = (uint8_t *)malloc(1); /* non-NULL for 0-length */
        *out_len = 0;
        return CC_OK;
    }

    /* Reconstruct block structure (same logic as encoder) */
    int data_per_block = CC_RS_MAX_BLOCK - 2;
    for (; data_per_block >= 1; data_per_block--) {
        int ec_count = cc_rs_ec_symbol_count(data_per_block, ec_ratio_100);
        if (data_per_block + ec_count <= CC_RS_MAX_BLOCK) break;
    }

    int num_blocks = ((int)payload_len + data_per_block - 1) / data_per_block;
    int *block_sizes = (int *)malloc((size_t)num_blocks * sizeof(int));
    if (!block_sizes) { cc_grid_free(&grid); free(pixels); return CC_ERR_ALLOC; }

    int total_stream = 0;
    for (int i = 0; i < num_blocks; i++) {
        int start = i * data_per_block;
        int end = start + data_per_block;
        if ((uint32_t)end > payload_len) end = (int)payload_len;
        int block_data_len = end - start;
        int block_ec = cc_rs_ec_symbol_count(block_data_len, ec_ratio_100);
        block_sizes[i] = block_data_len + block_ec;
        total_stream += block_sizes[i];
    }

    int total_stream_cells = (total_stream + bytes_per_cell - 1) / bytes_per_cell;

    /* Read data cells */
    int raw_stream_len = total_stream_cells * bytes_per_cell;
    uint8_t *raw_stream = (uint8_t *)calloc((size_t)raw_stream_len, 1);
    if (!raw_stream) {
        free(block_sizes);
        cc_grid_free(&grid);
        free(pixels);
        return CC_ERR_ALLOC;
    }

    for (int i = 0; i < total_stream_cells; i++) {
        int coord_idx = header_cells + i;
        if (coord_idx >= grid.data_count) break;
        int cx = grid.data_coords[coord_idx * 2];
        int cy = grid.data_coords[coord_idx * 2 + 1];
        const uint8_t *px = read_cell(pixels, pixel_width, bpp, cell_size,
                                       cx, cy);
        cc_decode_cell(px, mode, raw_stream + i * bytes_per_cell);
    }

    free(pixels);
    cc_grid_free(&grid);

    /* Deinterleave */
    uint8_t **rs_blocks = (uint8_t **)malloc((size_t)num_blocks * sizeof(uint8_t *));
    if (!rs_blocks) { free(raw_stream); free(block_sizes); return CC_ERR_ALLOC; }

    for (int i = 0; i < num_blocks; i++) {
        rs_blocks[i] = (uint8_t *)malloc((size_t)block_sizes[i]);
        if (!rs_blocks[i]) {
            for (int j = 0; j < i; j++) free(rs_blocks[j]);
            free(rs_blocks);
            free(raw_stream);
            free(block_sizes);
            return CC_ERR_ALLOC;
        }
    }

    cc_deinterleave(raw_stream, total_stream, rs_blocks, block_sizes,
                    num_blocks);
    free(raw_stream);

    /* RS decode each block */
    uint8_t *result = (uint8_t *)malloc((size_t)payload_len);
    if (!result) {
        for (int i = 0; i < num_blocks; i++) free(rs_blocks[i]);
        free(rs_blocks);
        free(block_sizes);
        return CC_ERR_ALLOC;
    }

    int result_offset = 0;
    for (int i = 0; i < num_blocks; i++) {
        int block_data_len = data_per_block;
        int remaining = (int)payload_len - i * data_per_block;
        if (block_data_len > remaining) block_data_len = remaining;
        int block_ec = block_sizes[i] - block_data_len;

        uint8_t *decoded = (uint8_t *)malloc((size_t)block_data_len);
        if (!decoded) {
            for (int j = 0; j < num_blocks; j++) free(rs_blocks[j]);
            free(rs_blocks); free(block_sizes); free(result);
            return CC_ERR_ALLOC;
        }

        rc = cc_rs_decode(rs_blocks[i], block_sizes[i], block_ec, decoded);
        if (rc != CC_OK) {
            free(decoded);
            for (int j = 0; j < num_blocks; j++) free(rs_blocks[j]);
            free(rs_blocks); free(block_sizes); free(result);
            return rc;
        }

        int copy_len = block_data_len;
        if (result_offset + copy_len > (int)payload_len)
            copy_len = (int)payload_len - result_offset;
        memcpy(result + result_offset, decoded, (size_t)copy_len);
        result_offset += copy_len;
        free(decoded);
    }

    for (int i = 0; i < num_blocks; i++) free(rs_blocks[i]);
    free(rs_blocks);
    free(block_sizes);

    /* Decompress if needed */
    if (header.compressed) {
        /* Estimate decompressed size — try progressively larger buffers */
        size_t decomp_cap = payload_len * 4;
        if (decomp_cap < 1024) decomp_cap = 1024;

        uint8_t *decompressed = NULL;
        int zrc;
        for (int attempt = 0; attempt < 10; attempt++) {
            decompressed = (uint8_t *)malloc(decomp_cap);
            if (!decompressed) { free(result); return CC_ERR_ALLOC; }

            uLongf dest_len = (uLongf)decomp_cap;
            zrc = uncompress(decompressed, &dest_len, result,
                             (uLong)payload_len);
            if (zrc == Z_OK) {
                free(result);
                *out_data = decompressed;
                *out_len = (size_t)dest_len;
                return CC_OK;
            }
            free(decompressed);
            if (zrc != Z_BUF_ERROR) {
                free(result);
                return CC_ERR_ZLIB;
            }
            decomp_cap *= 2;
        }

        free(result);
        return CC_ERR_ZLIB;
    }

    *out_data = result;
    *out_len = (size_t)payload_len;
    return CC_OK;
}
